diff --git a/vms/server/nx_vms_server/unit_tests/src/nx/mediaserver/fs/media_paths/media_paths_filter_ut.cpp b/vms/server/nx_vms_server/unit_tests/src/nx/mediaserver/fs/media_paths/media_paths_filter_ut.cpp
index e7408cb23cb..376f741eafc 100644
--- a/vms/server/nx_vms_server/unit_tests/src/nx/mediaserver/fs/media_paths/media_paths_filter_ut.cpp
+++ b/vms/server/nx_vms_server/unit_tests/src/nx/mediaserver/fs/media_paths/media_paths_filter_ut.cpp
@@ -80,7 +80,7 @@ protected:
 
     void whenPathsRequested()
     {
-        m_mediaPaths = media_paths::get(std::move(m_filterConfig));
+        m_mediaPaths = media_paths::getMediaPaths(std::move(m_filterConfig));
     }
 
     void thenNumberOfStoragesReturnedShouldBeEqualTo(int expected) const
